### PR TITLE
Compilation cleanup

### PR DESCRIPTION
**Pull request by graydon**
Some generic refactoring and tidying in the driver, mostly splitting `Compilation::performJobsImpl`
into enough pieces (and turning various lambdas and locals into named and commented methods
and members) that I can follow what it's doing. This is prep work for more involved debugging of
driver issues; I realized I couldn't really understand the structure of `performJobsImpl` before.

Along the way, fixed a few minor, mostly-logging issues:

  - Jobs are all now logged uniformly
  - Link and compile jobs look different now (this accounted for seemingly redundant log messages)
  - Skipped jobs are only logged and marked finished once, not once-per-taskqueue-iteration
  - MarkTracer is provided to all markTransitive calls, so it logs dependency-reasons more often
  - The confusing "" member provision case is logged as a special case
  - Failure to reload swiftdeps files when expected will emit a warning before rebuilding the world

rdar://30792302

### CODE REVIEW

File: lib/Driver/Compilation.cpp [hunk 1]
```diff
@@ -266,6 +266,18 @@ namespace driver {
         parseable_output::emitBeganMessage(llvm::errs(), *BeganCmd, Pid);
     }
 
+    void
+    dependencyLoadFailed(StringRef DependenciesFile, bool Warn=true) {
+      if (Warn)
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: I'm on the fence about this. On the one hand, we might get more feedback from users, but on the other they may end up with a warning they can't silence. (Neither Xcode nor SwiftPM expose a switch to turn off incremental builds.)

File: lib/Driver/Compilation.cpp [hunk 2]
```diff
@@ -590,8 +592,9 @@ namespace driver {
           }
 
           ScheduledCommands.insert(Cmd);
-          markFinished(Cmd);
+          markFinished(Cmd, /*Skipped=*/true);
         }
+        DeferredCommands.clear();
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: Ooh, good catch.

**Comment by jrose-apple** [comment]
I'm going to trust you that all the refactoring parts don't change anything; chopping up `performJobsImpl` is certainly not a bad thing. I think "PerformJobsState" may no longer be the right name for the thing you've created.

**Comment by graydon** [comment]
Pushed a couple updates to dial back the diagnostic chatter.

**Comment by graydon** [comment]
@swift-ci please test

**Comment by graydon** [comment]
@swift-ci please test and merge

**Comment by graydon** [comment]
@swift-ci please smoke test and merge

**Comment by graydon** [comment]
@swift-ci please smoke test

**Pull request merged**